File: src-tauri/src/models/connection.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::time::Duration;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ConnectionConfig {
    pub id: String,
    pub name: String,
    pub connection_type: ConnectionType,
    pub serial_config: Option<SerialConfig>,
    pub tcp_config: Option<TcpConfig>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum ConnectionType {
    Serial,
    Tcp,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SerialConfig {
    pub port: String,
    pub baud_rate: u32,
    pub data_bits: DataBits,
    pub stop_bits: StopBits,
    pub parity: Parity,
    pub flow_control: FlowControl,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TcpConfig {
    pub host: String,
    pub port: u16,
    #[serde(with = "duration_serde")]
    pub timeout: Duration,
    pub keep_alive: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum DataBits {
    Five,
    Six,
    Seven,
    Eight,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum StopBits {
    One,
    OnePointFive,
    Two,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum Parity {
    None,
    Even,
    Odd,
    Mark,
    Space,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum FlowControl {
    None,
    Software,
    Hardware,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum ConnectionStatus {
    Disconnected,
    Connecting,
    Connected,
    Error,
}

// Duration serialization helper
mod duration_serde {
    use serde::{Deserialize, Deserializer, Serialize, Serializer};
    use std::time::Duration;

    pub fn serialize<S>(duration: &Duration, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        duration.as_millis().serialize(serializer)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Duration, D::Error>
    where
        D: Deserializer<'de>,
    {
        let millis = u64::deserialize(deserializer)?;
        Ok(Duration::from_millis(millis))
    }
}

impl Default for SerialConfig {
    fn default() -> Self {
        Self {
            port: String::new(),
            baud_rate: 115200,
            data_bits: DataBits::Eight,
            stop_bits: StopBits::One,
            parity: Parity::None,
            flow_control: FlowControl::None,
        }
    }
}

impl Default for TcpConfig {
    fn default() -> Self {
        Self {
            host: "127.0.0.1".to_string(),
            port: 8080,
            timeout: Duration::from_secs(5),
            keep_alive: true,
        }
    }
}

impl ConnectionConfig {
    pub fn new_serial(name: String, serial_config: SerialConfig) -> Self {
        let now = Utc::now();
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name,
            connection_type: ConnectionType::Serial,
            serial_config: Some(serial_config),
            tcp_config: None,
            created_at: now,
            updated_at: now,
        }
    }

    pub fn new_tcp(name: String, tcp_config: TcpConfig) -> Self {
        let now = Utc::now();
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name,
            connection_type: ConnectionType::Tcp,
            serial_config: None,
            tcp_config: Some(tcp_config),
            created_at: now,
            updated_at: now,
        }
    }
}

// Convert between our types and serialport types
impl From<DataBits> for serialport::DataBits {
    fn from(value: DataBits) -> Self {
        match value {
            DataBits::Five => serialport::DataBits::Five,
            DataBits::Six => serialport::DataBits::Six,
            DataBits::Seven => serialport::DataBits::Seven,
            DataBits::Eight => serialport::DataBits::Eight,
        }
    }
}

impl From<StopBits> for serialport::StopBits {
    fn from(value: StopBits) -> Self {
        match value {
            StopBits::One => serialport::StopBits::One,
            StopBits::OnePointFive => serialport::StopBits::Two, // Note: serialport crate doesn't have 1.5
            StopBits::Two => serialport::StopBits::Two,
        }
    }
}

impl From<Parity> for serialport::Parity {
    fn from(value: Parity) -> Self {
        match value {
            Parity::None => serialport::Parity::None,
            Parity::Even => serialport::Parity::Even,
            Parity::Odd => serialport::Parity::Odd,
            Parity::Mark => serialport::Parity::None, // Note: serialport crate doesn't have Mark/Space
            Parity::Space => serialport::Parity::None,
        }
    }
}

impl From<FlowControl> for serialport::FlowControl {
    fn from(value: FlowControl) -> Self {
        match value {
            FlowControl::None => serialport::FlowControl::None,
            FlowControl::Software => serialport::FlowControl::Software,
            FlowControl::Hardware => serialport::FlowControl::Hardware,
        }
    }
}

File: src/lib/types/connection.ts
export interface ConnectionConfig {
  id: string;
  name: string;
  connection_type: ConnectionType;
  serial_config?: SerialConfig;
  tcp_config?: TcpConfig;
  created_at: string;
  updated_at: string;
}

export enum ConnectionType {
  Serial = 'Serial',
  Tcp = 'Tcp',
}

export interface SerialConfig {
  port: string;
  baud_rate: number;
  data_bits: DataBits;
  stop_bits: StopBits;
  parity: Parity;
  flow_control: FlowControl;
}

export interface TcpConfig {
  host: string;
  port: number;
  timeout: number; // milliseconds
  keep_alive: boolean;
}

export enum DataBits {
  Five = 'Five',
  Six = 'Six',
  Seven = 'Seven',
  Eight = 'Eight',
}

export enum StopBits {
  One = 'One',
  OnePointFive = 'OnePointFive',
  Two = 'Two',
}

export enum Parity {
  None = 'None',
  Even = 'Even',
  Odd = 'Odd',
  Mark = 'Mark',
  Space = 'Space',
}

export enum FlowControl {
  None = 'None',
  Software = 'Software',
  Hardware = 'Hardware',
}

export enum ConnectionStatus {
  Disconnected = 'Disconnected',
  Connecting = 'Connecting',
  Connected = 'Connected',
  Error = 'Error',
}

// UI用の選択肢オプション
export interface SelectOption<T = string> {
  value: T;
  label: string;
  disabled?: boolean;
}

// 接続関連のUI状態
export interface ConnectionState {
  currentConnection: ConnectionConfig | null;
  connectionStatus: ConnectionStatus;
  availablePorts: string[];
  profiles: ConnectionConfig[];
  isLoading: boolean;
  error: string | null;
}

// シリアルポート情報
export interface SerialPortInfo {
  port_name: string;
  port_type?: string;
  vid?: number;
  pid?: number;
  serial_number?: string;
  manufacturer?: string;
  product?: string;
}

// 定数
export const BAUD_RATES = [
  1200, 2400, 4800, 9600, 19200, 38400, 57600, 115200, 230400, 460800, 921600,
] as const;

export const DATA_BITS_OPTIONS: SelectOption<DataBits>[] = [
  { value: DataBits.Five, label: '5' },
  { value: DataBits.Six, label: '6' },
  { value: DataBits.Seven, label: '7' },
  { value: DataBits.Eight, label: '8' },
];

export const STOP_BITS_OPTIONS: SelectOption<StopBits>[] = [
  { value: StopBits.One, label: '1' },
  { value: StopBits.OnePointFive, label: '1.5' },
  { value: StopBits.Two, label: '2' },
];

export const PARITY_OPTIONS: SelectOption<Parity>[] = [
  { value: Parity.None, label: 'なし' },
  { value: Parity.Even, label: '偶数' },
  { value: Parity.Odd, label: '奇数' },
  { value: Parity.Mark, label: 'マーク' },
  { value: Parity.Space, label: 'スペース' },
];

export const FLOW_CONTROL_OPTIONS: SelectOption<FlowControl>[] = [
  { value: FlowControl.None, label: 'なし' },
  { value: FlowControl.Software, label: 'ソフトウェア' },
  { value: FlowControl.Hardware, label: 'ハードウェア' },
];

// ヘルパー関数
export function createDefaultSerialConfig(): SerialConfig {
  return {
    port: '',
    baud_rate: 115200,
    data_bits: DataBits.Eight,
    stop_bits: StopBits.One,
    parity: Parity.None,
    flow_control: FlowControl.None,
  };
}

export function createDefaultTcpConfig(): TcpConfig {
  return {
    host: '127.0.0.1',
    port: 8080,
    timeout: 5000,
    keep_alive: true,
  };
}

export function createNewConnectionConfig(
  name: string,
  type: ConnectionType,
  config: SerialConfig | TcpConfig
): Omit<ConnectionConfig, 'id' | 'created_at' | 'updated_at'> {
  const now = new Date().toISOString();
  return {
    name,
    connection_type: type,
    ...(type === ConnectionType.Serial
      ? { serial_config: config as SerialConfig }
      : { tcp_config: config as TcpConfig }),
  };
}

export function getConnectionDisplayName(config: ConnectionConfig): string {
  if (config.connection_type === ConnectionType.Serial && config.serial_config) {
    return `${config.name} (${config.serial_config.port})`;
  } else if (config.connection_type === ConnectionType.Tcp && config.tcp_config) {
    return `${config.name} (${config.tcp_config.host}:${config.tcp_config.port})`;
  }
  return config.name;
}

export function validateSerialConfig(config: SerialConfig): string[] {
  const errors: string[] = [];
  
  if (!config.port.trim()) {
    errors.push('ポートを選択してください');
  }
  
  if (config.baud_rate <= 0) {
    errors.push('有効なボーレートを入力してください');
  }
  
  return errors;
}

export function validateTcpConfig(config: TcpConfig): string[] {
  const errors: string[] = [];
  
  if (!config.host.trim()) {
    errors.push('ホストアドレスを入力してください');
  }
  
  if (config.port <= 0 || config.port > 65535) {
    errors.push('有効なポート番号（1-65535）を入力してください');
  }
  
  if (config.timeout <= 0) {
    errors.push('有効なタイムアウト値を入力してください');
  }
  
  return errors;
}